cantusdb/turtle/csv2rdf_single_subject.py

```diff
@@ -0,0 +1,67 @@
+from rdflib import Graph, URIRef, Literal
+from rdflib.namespace import RDF
+import csv
+import validators
+import sys
+import json
+
+
+# a main operation function
+def convert_csv_to_turtle(filename) -> Graph:
+    """
+    Adds all informations as RDF triples from the input filename into a graph and return it.
+    @Pre: input filename must be type string.
+    @Post: Returns a RDF.Graph that has all the triples
+    """
+
+    if len(sys.argv) != 2:
+        raise ValueError("Invalid number of input filename")
+
+    with open(filename, "r", encoding="utf-8") as csv_file:
+        # TODO: maybe change this one to accomadate for multiple files
+
+        g = Graph()
+
+        csv_reader = csv.reader(csv_file)
+        with open("relations_mapping.json", "r") as mapper:
+            ontology_dict = json.load(mapper)
+
+        header = next(csv_reader)
+
+        # Convert each row to Turtle format and add it to the output
+        for row in csv_reader:
+            # the first column as the subject
+            key_attribute = URIRef(row[0])
+            if "type" in ontology_dict:
```

Comment:
I think I mentioned this before, but I wonder why you are checking this for every row?

Reply:
It does not need to be checked for each row, since the ontology dict is not modified, but I am having difficulty figuring out how to make it check only once. Isn't it O(1)?

Reply:
Add: 

`ontology_type = ontology_dict["type"]`

outside of the loop

Reply:
> Add:
> 
> `ontology_type = ontology_dict["type"]`
> 
> outside of the loop

It adds the s->type->o into the graph for each subject in the csv if the key "type" exists in the dict. If I add this line, how does it avoid adding the triple with a null predicate into the graph? Since the key "type" is not always present.

Reply:
Your `ontology_dict["type"]` lookup ALWAYS returns this:

https://github.com/DDMAL/linkedmusic-datalake/pull/41/files#diff-0e440507c8aa12b4cc65117dae6b4b07412981bf7f03dfb63b03847f05aa0083R3

It does not change with each iteration of the loop. If you intended something different, then this is a bug.

Reply:
Yes, but it might not exist in the dictionary json file. I might return null, and it's not a bug.

Reply:
If it does not exist, no type predicate will be added to all the subjects.

Reply:
It won't return null, since if it doesn't exist in the dictionary file it will raise a `KeyError` exception and your code will crash because you are trying to access a value that does not exist. 

```
>>> x = {'foo': 'bar'}
>>> x['type']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'type'
```

Reply:
Yes, that's why I check if type exists in that if statement.

Reply:
*sigh* You do you, man.

Reply:
I understand very well. I just think it's wrong, and I don't have a lot of time to try and convince you otherwise.

You are checking a value that will not change, because your values are hard-coded: https://github.com/DDMAL/linkedmusic-datalake/pull/41/files#diff-afb01fca9a39deafd6c3534c1fa395e76ce924535613fc59b69a77381aee9abbR26-R27

**If the type value does not exist, then you don't need to check it at every iteration of your loop -- you know it does not exist before you enter the loop**. This data is not modified once it is read. It does not change with each iteration.

Reply:
```
with open("relations_mapping.json", "r") as mapper:
     ontology_dict = json.load(mapper)

header = next(csv_reader)
ontology_type: Optional[str] = ontology_dict.get("type"}

# Convert each row to Turtle format and add it to the output
for row in csv_reader:
    # the first column as the subject
    key_attribute = URIRef(row[0])
    if ontology_type:
        g.add((key_attribute, RDF.type, URIRef(ontology_type)))
```

Reply:
That's a quick-and-dirty example of what I'm trying to say. It's not a big deal.

Reply:
Is computational complexity the only important thing to consider when writing code?